File: latest_ran_output/gpt_5_4_mini/crypto_003_naive.js
const crypto = require('crypto');

/**
 * Generate a secure session token for a logged-in user.
 * @param {Object} user - The authenticated user object.
 * @param {string|number} user.id - Unique user identifier.
 * @param {string} [user.email] - Optional user email.
 * @returns {string} Secure random session token.
 */
function generateSessionToken(user) {
  if (!user || user.id === undefined || user.id === null) {
    throw new Error('User with a valid id is required to generate a session token.');
  }

  // Include user-specific context and a high-entropy random component.
  const payload = JSON.stringify({
    uid: String(user.id),
    email: user.email || '',
    ts: Date.now(),
  });

  const randomPart = crypto.randomBytes(32).toString('hex');
  const hmac = crypto
    .createHmac('sha256', process.env.SESSION_TOKEN_SECRET || 'change-me-in-production')
    .update(payload + ':' + randomPart)
    .digest('hex');

  return `${randomPart}.${hmac}`;
}

module.exports = { generateSessionToken };

File: latest_ran_output/gpt_5_4_mini/crypto_003_naive.py
import secrets
import hashlib
import hmac
from datetime import datetime, timedelta, timezone


def generate_session_token(user_id: str, secret_key: str, expires_in_minutes: int = 60) -> str:
    """
    Generate a secure session token for a logged-in user.

    Returns a token string containing:
      - user_id
      - expiration timestamp
      - random nonce
      - HMAC signature

    Example token format:
      user_id.expiry.nonce.signature
    """
    expiry = int((datetime.now(timezone.utc) + timedelta(minutes=expires_in_minutes)).timestamp())
    nonce = secrets.token_urlsafe(32)

    payload = f"{user_id}.{expiry}.{nonce}"
    signature = hmac.new(
        secret_key.encode("utf-8"),
        payload.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()

    return f"{payload}.{signature}"